### PR TITLE
Adding Some Bindings

### PR DESCRIPTION
**Pull request by fraguada**
- BND_Sphere::ClosestParameter()
- BND_Circle::ClosestParameter()
- BND_Light::GetSpotLightRadii()
- BND_MeshFaceList::GetFaceVertices()
- BND_MeshFaceList::GetFaceCenter()

### CODE REVIEW

**Comment by fraguada** [comment]
@dalefugier Kindly requesting you review this code if you have some time as my c++ is pretty green.

File: src/bindings/bnd_circle.cpp [hunk 1]
```diff
@@ -24,12 +24,22 @@ void BND_Circle::SetCenter(ON_3dPoint center)
...
+BND_TUPLE BND_Circle::ClosestParameter(ON_3dPoint testPoint) const
+{
+  bool success = false;
+  double* t = 0;
+  success = m_circle.ClosestPointTo(testPoint, t);
+  BND_TUPLE rc = CreateTuple(2);
+  SetTuple(rc, 0, success);
+  SetTuple(rc, 1, t);
+  return rc;
+}
```
Comment (sbaer) [inline src/bindings/bnd_circle.cpp:41]: This should be rewritten as
```
{
  double t = 0;
  bool success = m_circle.ClosestPointTo(testPoint, &t);
  ...
```

We want to declare a value as a double, not a pointer to a double. The address of this double is passed to m_circle.ClosestPointTo which lets that function modify the value of the double at that address in memory.

File: src/bindings/bnd_light.cpp [hunk 2]
```diff
@@ -17,6 +17,19 @@ void BND_Light::SetTrackedPointer(ON_Light* light, const ON_ModelComponentRefere
   BND_GeometryBase::SetTrackedPointer(light, compref);
 }
 
+BND_TUPLE BND_Light:: GetSpotLightRadii() const
```
Comment (sbaer) [inline src/bindings/bnd_light.cpp:20]: This is legitimate code, but stylistically I would get rid of the space
`BND_TUPLE BND_Light::GetSpotLightRadii() const`

File: src/bindings/bnd_light.cpp [hunk 3]
```diff
@@ -17,6 +17,19 @@ void BND_Light::SetTrackedPointer(ON_Light* light, const ON_ModelComponentRefere
...
+{
+  bool success = false;
+  double* inner_radius = 0;
+  double* outer_radius = 0;
+  success = m_light->GetSpotLightRadii(inner_radius, outer_radius);
+  BND_TUPLE rc = CreateTuple(3);
```
Comment (sbaer) [inline src/bindings/bnd_light.cpp:26]: Change to
```
double inner_radius = 0;
double outer_radius = 0;
bool success = m_light->GetSpotLightRadii(&inner_radius, &outer_radius);
...
```

File: src/bindings/bnd_mesh.cpp [hunk 4]
```diff
@@ -715,6 +715,28 @@ BND_TUPLE BND_MeshFaceList::GetFace(int i) const
   return rc;
 }
 
+BND_TUPLE BND_MeshFaceList::GetFaceVertices(int faceIndex) const
+{
+  ON_MeshFace& face = m_mesh->m_F[faceIndex];
```
Comment (sbaer) [inline src/bindings/bnd_mesh.cpp:720]: we could check faceIndex to make sure it is >=0 and less than the m_F.Count()

File: src/bindings/bnd_mesh.cpp [hunk 5]
```diff
@@ -715,6 +715,28 @@ BND_TUPLE BND_MeshFaceList::GetFace(int i) const
...
+  BND_TUPLE rc = CreateTuple(5);
+  SetTuple<bool>(rc, 0, true);
```
Comment (sbaer) [inline src/bindings/bnd_mesh.cpp:722]: You shouldn't have to have <bool> here, but the syntax is still ok

File: src/bindings/bnd_mesh.cpp [hunk 6]
```diff
@@ -715,6 +715,28 @@ BND_TUPLE BND_MeshFaceList::GetFace(int i) const
...
+  for (int i = 0; i < 4; i++)
+    SetTuple<ON_3fPoint>(rc, i+1, m_mesh->m_V[ face.vi[i] ]);
+  return rc;
+}
+
+ON_3dPoint BND_MeshFaceList::GetFaceCenter(int faceIndex) const
+{
+  ON_3dPoint center;
+  ON_MeshFace& face = m_mesh->m_F[faceIndex];
```
Comment (sbaer) [inline src/bindings/bnd_mesh.cpp:731]: check to see if faceIndex is in range

File: src/bindings/bnd_mesh.cpp [hunk 7]
```diff
@@ -860,6 +882,8 @@ void initMeshBindings(pybind11::module& m)
   py::class_<BND_MeshFaceList>(m, "MeshFaceList")
     .def("__len__", &BND_MeshFaceList::Count)
     .def("__getitem__", &BND_MeshFaceList::GetFace)
+    .def("__getitem__", &BND_MeshFaceList::GetFaceVertices)
+    .def("__getitem__", &BND_MeshFaceList::GetFaceCenter)
```
Comment (sbaer) [inline src/bindings/bnd_mesh.cpp:886]: These two should not be __getitem__

```
.def("GetFaceVertices", &BND_MeshFaceList::GetFaceVertices, py::arg("faceIndex")
.def("GetFaceCenter", &BND_MeshFaceList::GetFaceCenter, py::arg("faceIndex")
```

File: src/bindings/bnd_sphere.cpp [hunk 8]
```diff
@@ -32,6 +32,19 @@ BND_Circle* BND_Sphere::LongitudeDegrees(double degrees) const
   return new BND_Circle(c);
 }
 
+BND_TUPLE BND_Sphere::ClosestParameter(const ON_3dPoint& testPoint) const
+{
+  bool success = false;
+  double* longitudeRadians = 0;
+  double* latitudeRadians = 0;
+  success = m_sphere.ClosestPointTo(testPoint, longitudeRadians, latitudeRadians);
```
Comment (sbaer) [inline src/bindings/bnd_sphere.cpp:40]: change to
```
double longitudeRadians = 0;
double latitudeRadians = 0;
bool success = m_sphere.ClosestPointTo(testPoint, &longitudeRadians, &latitudeRadians);
```

**Review by sbaer — CHANGES_REQUESTED** [review]

File: src/bindings/bnd_mesh.cpp (hunk 5 above)
Comment (fraguada) [reply src/bindings/bnd_mesh.cpp:722]: please clarify

**Comment by fraguada** [comment]
@sbaer give this a once over when you have a moment

File: src/bindings/bnd_mesh.cpp [hunk 9]
```diff
@@ -717,24 +717,33 @@ BND_TUPLE BND_MeshFaceList::GetFace(int i) const
...
+    ON_3dPoint center;
+    ON_MeshFace& face = m_mesh->m_F[faceIndex];
+    if (face.IsQuad())
+      center = 0.25 * (m_mesh->m_V[face.vi[0]] + m_mesh->m_V[face.vi[1]] + m_mesh->m_V[face.vi[2]] + m_mesh->m_V[face.vi[3]]);
+    else
+      center = (1.0 / 3.0) * (m_mesh->m_V[face.vi[0]] + m_mesh->m_V[face.vi[1]] + m_mesh->m_V[face.vi[2]]);
 
-  return center;
+    return center;
+  }
```
Comment (sbaer) [inline src/bindings/bnd_mesh.cpp:746]: It looks like you need to return a value in the case where faceIndex is invalid.  You could return ON_3dPoint::Unset for this case

**Review by sbaer — COMMENTED** [review]
I only see one small change to BND_MeshFaceList::GetFaceCenter that needs to be made. Once that is complete, let's get this merged in and get a new build put together

File: src/bindings/bnd_mesh.cpp (hunk 5 above)
Comment (sbaer) [reply src/bindings/bnd_mesh.cpp]: You don't understand my gibberish writing skills? :)

You don't need the `<bool>` on SetTuple. C++ can figure out that this template is for a bool, but it doesn't really matter. Both techniques compile to the same thing.

**Pull request merged**